pilot/model/controller_view.go

```diff
@@ -0,0 +1,580 @@
+		if old.Endpoint.ServicePort.Name != "" {
+			cv.instanceLabels.deleteLabel(k, labelInstanceNamedPort, old.Endpoint.ServicePort.Name)
+		}
+		for label, value := range old.Labels {
+			cv.instanceLabels.deleteLabel(k, label, value)
+		}
+		delete(cv.serviceInstances, k)
+	}
+	if e != EventDelete {
+		// Treat as upsert
+		cv.instanceLabels.addLabel(k, labelServiceName, i.Service.Hostname)
+		cv.instanceLabels.addLabel(k, labelInstanceIP, getNormalizedIP(i.Endpoint.Address))
+		cv.instanceLabels.addLabel(k, labelInstancePort, getNormalizedPort(i.Endpoint.Port))
+		if i.Endpoint.ServicePort != nil && i.Endpoint.ServicePort.Name != "" {
+			cv.instanceLabels.addLabel(k, labelInstanceNamedPort, i.Endpoint.ServicePort.Name)
+		}
+		for label, value := range i.Labels {
+			cv.instanceLabels.addLabel(k, label, value)
+		}
+		cv.serviceInstances[k] = i
+	}
+}
+
+// Given a set of labels, gets a set of keys that match the values
+// of all labels
+func (vl viewLabels) getKeysMatching(labels Labels) viewKeySet {
+	type matchingSet struct {
+		labelName  string
+		labelValue string
+		keySet     viewKeySet
+	}
+	countLabels := len(labels)
+	if countLabels == 0 {
+		// There must be at least one label else return nothing
+		return viewKeySet{}
+	}
+	// Note: 0th index has the smallest keySet
+	matchingSets := make([]matchingSet, countLabels)
+	smallestSetLen := math.MaxInt32
+	setIdx := 0
+	for l, v := range labels {
+		valueKeysetMap := vl[l]
+		if len(valueKeysetMap) == 0 {
+			// Nothing matched at least one label name
+			return viewKeySet{}
+		}
+		valKeySet := valueKeysetMap[v]
+		lenKeySet := len(valKeySet)
+		if lenKeySet == 0 {
+			// There were no service keys for this label value
+			return viewKeySet{}
+		}
+		matchingSets[setIdx].keySet = valKeySet
+		if lenKeySet < smallestSetLen {
+			smallestSetLen = lenKeySet
+			if setIdx > 0 {
+				swappedMatchingSet := matchingSets[0]
+				matchingSets[0] = matchingSets[setIdx]
+				matchingSets[setIdx] = swappedMatchingSet
+			}
+		}
+		setIdx++
+	}
+	if countLabels == 1 {
+		return matchingSets[0].keySet
+	}
+	out := newViewKeySet(matchingSets[0].keySet)
+	for k := range out {
+		for setIdx := 1; setIdx < countLabels; setIdx++ {
+			_, found := matchingSets[setIdx].keySet[k]
+			if !found {
+				delete(out, k)
+				break
+			}
+		}
+	}
+	return out
+}
+
+// Given a label name, fetches a set of keys that have the label,
+// irrespective of what the label value may be.
+func (vl viewLabels) getAllKeysMatchingLabel(l string) viewKeySet {
+	valueKeysetMap := vl[l]
+	if len(valueKeysetMap) == 0 {
+		// Nothing matched at least one label name
+		return viewKeySet{}
+	}
+	out := viewKeySet{}
+	for _, keySet := range valueKeysetMap {
+		out.appendAll(keySet)
+	}
+	return out
+}
+
+// addLabel creates the reverse lookup by label name and label value for the key k.
+func (vl viewLabels) addLabel(k, labelName, labelValue string) {
+	valueKeysetMap, labelNameFound := vl[labelName]
```

Comment:
here and elsewhere, `labelNameFound` -> `exists`?
(entirely subjective suggestion --feel free to ignore)